### ISSUE TITLE
Jbrowse - Taking too long to load and link issues.

### ISSUE THREAD

**Issue by Paul-Davis** [WormBase/website#4740]
Hi,

Was trying to sort something for a user and wanted to suggest that they explore the Jbrowse functionality, but Jbrowse seems to be taking 5min plus to load even on repeat sessions which isn't usable and when it does load, the region is not what the gene page specified in the text link co-ordinates.

It's actually making firefox unusable on my machine

Paul

**Comment by Paul-Davis** [comment]
After additional testing, it seems that its the "Full Page" jbrowse entry points that take a very long time with the main location link not dislaying the stated co-ordinates.

![4740_1](https://cloud.githubusercontent.com/assets/717707/14457759/cbc94ad4-00a4-11e6-9983-98780b1d1ad1.png)

Additional options in the working embedded jbrowse

![4740_2](https://cloud.githubusercontent.com/assets/717707/14457771/e386415e-00a4-11e6-9683-fb42b07c1156.png)

**Comment by scottcain** [comment]
Hi @Paul-Davis Thanks for the report.  I was unaware of the problem you're reporting with the coordinates on the gene pages not being correct; I'll look into that.  Also, I didn't realize the that the "second" full page button had returned; I removed it once but must have inadvertently reintroduced it. Unfortunately, the button you like (the nice and speedy one) is the one I want to get rid of!

The JBrowse for C. elegans is very slow to load due to the enormous number of modENCODE tracks.  I didn't realize that it was now completely killing Firefox (I killed it after 6 minutes), whereas Safari took 4 minutes and Chrome took 1.5 minutes.  The "nice and speedy" JBrowse is a stripped down instance with just 30 or so tracks created just for gene pages (it's not available from the JBrowse pulldown menu, and when you clicked on the full-screen view, it probably reported that you were looking at B. malayi).  In the past, I suggested creating two separate JBrowse instances for C. elegans, with and without modENCODE to give users that don't care about modENCODE data a good experience, but that idea shot down at the time but I suspect it is time to revisit it.

**Comment by khowe** [comment]
This is another use-case for TrackHubs. If JBrowse supported them, we could pull out the modENCODE tracks into a Hub. This would giving fast performance for people not interested in modENCODE, but the ability to connect the modENCODE tracks on-demand. 

Who can we lobby to get TrackHub support in JBrowse?

**Comment by scottcain** [comment]
Even though I know I've heard the developers discuss it, I created an issue to poke them in public about it:

https://github.com/GMOD/jbrowse/issues/728

**Comment by Paul-Davis** [comment]
Thanks @scottcain,

I think I don't fully get the different views of Jbrowse on WormBase.

Which version of the WormBase Jbrowse instillation is this: http://www.wormbase.org/tools/genome//jbrowse/full.html?overview=1&loc=I%3A14390386..14396207&data=data%2Fc_elegans_simple&nav=1&tracks=Classical_alleles%2CPolymorphisms%2CCurated_Genes&highlight=&tracklist=1

I like this one but as you say it doesn't have modENCODE....could this be used as the default view with modENCODE a separate option with the expected delay because of all the tracks?

**Comment by scottcain** [comment]
That is a jbrowse instance created explicitly for making fast views on gene pages.  If I were to implement a non-modENCODE JBrowse instance, I'd add more tracks, to a total of 60 or so, but it would still be very fast.  I'm not against doing that, but the last time I discussed it with the web team, @tharris didn't like the idea too much.  I think he may reconsider now that it's killing Firefox :-/

From a UI perspective, this is what I'd do:
1. Remove modENODE tracks from the Hawaiian strain browser completely (You may not have noticed they're there)
2. Create a JBrowse-slim for C. elegans with all of the non-modENCODE tracks (around 60) and refer to it just as C. elegans in the UI, and this would become the default JBrowse linked to from the Tools menu and in gene pages.
3. Rename the existing C. elegans instance to "C. elegans w/modENCODE" where it would be available from the dropdown menu in JBrowse.  In this state, it would be relatively difficult for the typical user to discover (which is probably why @tharris didn't like this solution when we last discussed it).  I don't have a good solution for this discoverability problem.

**Comment by khowe** [comment]
Hi @scottcain , I've only just noticed that the modENCODE tracks for the Hawaiian strain genome, both on Gbrowse and Jbrowse! 

The thing is, the Hawaiian strain genome is a different sequence, with a different coordinate system. It's technically not correct to be show tracks generated w.r.t. the N2 reference on this gene, because the features will not be in the correct place. 

I supposed it could be argued that the modENCODE tracks are still useful because the features will mostly be in approximately the right place. Was there a discussion that I missed that arrived at this conclusion?

**Comment by tharris** [comment]
I have *no* problem creating slim jbrowse instances. I certainly wouldn't rename the primary, default C. elegans JBrowse, just call the slim versions "C. elegans, no modENOCDE" or something. We could even hide it from the main selection screen if we just want to use it for driving inline displays.

Slightly disapointing that JBrowse, as the successor to GBrowse, seems to actually perform worse for things and have bigger management overhead.

@khowe as for CB4856, no, the tracks were just added by default. Creating the configuration files is done programmatically b y reading the contents of the species/ directory.

**Comment by scottcain** [comment]
The Hawaiian strain stuff with modENCODE was done automatically, and I thought it was little goofy, but since the modENCODE data don't actually map correctly to PRJNA13758 either, I figured the extra data could be useful when the mapping is roughly correct.  The significant downside of including it though makes the effort of manually removing them seem worth it.

@tharris what you're describing for the PRJNA13758 is essentially what we have now: a C. elegans slim that is (imperfectly at the moment) hidden from users except for driving displays.  What I'd suggest is that we make it more visible and allow users to go from the gene page to the slim version by default, as well as making it the default landing place when getting there from the Tools menu.

I agree with your frustration over JBrowse's slowness with so many tracks.  In the past, the Rat Genome Database (http://rgd.mcw.edu) is the site that pushed GBrowse with respect to numbers of tracks (at one point they had over 1000 in GBrowse) but now they have only about 100 tracks in both their GBrowse and JBrowse.  I suspect nobody comes close to having as many tracks in JBrowse as we are trying to do.  I did provide a test system to the JBrowse developers when I met with them in January to demonstrate the magnitude of the slowness problem to help prompt them into coming up with a solution.

**Comment by sibyl229** [comment]
Hi @Paul-Davis @scottcain about the link appearing incorrect, I think it's probably because we programmatically create a wider range based on the actual range of the gene. (for example, if the gene coordinate is start: 1000, end: 2000, then the range it shows will be something like start: 500, end: 2500, I don't remember the exact formula for padding.) This allows the gene to appear in the center of the JBrowse window, with its sorrounding features shown on both sides of the gene. Otherwise, the gene stretches across the window, and the user would have to zoom out to see anything sorrounding the gene.

**Comment by Paul-Davis** [comment]
@sibyl229 yes I remember the padding now, but thought it was much bigger when the jbrowse finally opened, but maybe that was an issue my end....padding to center the gene is fine and what the user would expect...thanks for the clarification.

**Comment by scottcain** [comment]
The main jbrowse instance on both staging and production are broken right now, but I did manage to speed it up by cutting down dramatically on the amount of metadata associated with each of the modencode tracks.  Take a look at 

http://staging.wormbase.org/tools/genome/jbrowse/full.html?data=data%2Fc_elegans_PRJNA13758

(this page bypasses the wormbase banner and javascript which is the source of the current problem)

To get a sense of the improved loading speed.  I'll continue to work on the metadata, which seems to be the source of the main slow down.
